[flang][OpenMP] Sema checks, lowering with new format of MAP modifiers

kparzysz opened the pull request:
OpenMP 6.0 has changed the modifiers on the MAP clause. Previous patch has introduced parsing support for them. This patch introduces processing of the new forms in semantic checks and in lowering. This only applies to existing modifiers, which were updated in the 6.0 spec. Any of the newly introduced modifiers (SELF and REF) are ignored.

kparzysz committed:
[flang][OpenMP] Parse OpenMP 6.0 map modifiers

OpenMP 6.0 has changed the modifiers on the MAP clause:
- map-type-modifier has been split into individual modifiers,
- map-type "delete" has become a modifier,
- new modifiers have been added.

This patch adds parsing support for all of the OpenMP 6.0 modifiers.
The old "map-type-modifier" is retained, but is no longer created in
parsing. It will remain to take advantage of the preexisting modifier
validation for older versions: when the OpenMP version is < 6.0,
the modifiers will be rewritten back as map-type-modifiers (or map-
type in case of "delete").

In this patch the modifiers will always be rewritten in the older
format to isolate these changes to parsing as much as possible.

kparzysz committed:
[flang][OpenMP] Sema checks, lowering with new format of MAP modifiers

OpenMP 6.0 has changed the modifiers on the MAP clause. Previous patch
has introduced parsing support for them. This patch introduces processing
of the new forms in semantic checks and in lowering. This only applies
to existing modifiers, which were updated in the 6.0 spec. Any of the
newly introduced modifiers (SELF and REF) are ignored.

kparzysz committed:
Simplify converting MapType, NFC

kparzysz committed:
Fix unit test

kparzysz commented:
Stack:
- https://github.com/llvm/llvm-project/pull/149134
- https://github.com/llvm/llvm-project/pull/149137 (this PR)

tblah reviewed (commented):
Please could you add a lit tests for lowering: one showing OpenMP 5 style modifiers being converted to the new ones, and one showing the new ones working in OpenMP 6.0.

kparzysz committed:
Merge branch 'main' into users/kparzysz/m01-map60-parser

kparzysz committed:
Don't pass SemanticsContext to CanonicalizeOmp yet

kparzysz committed:
Merge branch 'users/kparzysz/m01-map60-parser' into users/kparzysz/m02-map60-repr

kparzysz committed:
Pass SemanticsContext to CanonicalizeOmp

kparzysz committed:
Invent OmpMapStorage for map flags

We're using "storage" instead of "alloc"/"release" in lowering, this will
make things more consistent.

kparzysz committed:
Add testcase for using v60 modifiers in v52

kparzysz committed:
Clearer names for enum-converting functions for map modifiers

kparzysz commented on flang/lib/Lower/OpenMP/Clauses.cpp:
```diff
@@ -1003,12 +1003,13 @@ Map make(const parser::OmpClause::Map &inp,
   CLAUSET_ENUM_CONVERT( //
       convert1, parser::OmpMapType::Value, Map::MapType,
```
Done

kparzysz committed:
Remove now unused isMapEnteringConstruct

tblah commented:
> There is a parsing test that shows the conversion, plus parsing of the new modifiers (in the parsing PR). Do you want something specific to lowering, e.g. show the same MLIR generated for a given modifier with -fopenmp-version=52 and 60?

Ahh okay. Maybe we only need one test then, but it would be good to make sure this makes it through the clause processor okay.

kparzysz committed:
Merge branch 'main' into users/kparzysz/m02-map60-repr

kparzysz committed:
Check translation of map-type-modifiers

kparzysz commented:
> Ahh okay. Maybe we only need one test then, but it would be good to make sure this makes it through the clause processor okay.

I added more RUN lines to flang/test/Lower/OpenMP/map-modifiers.f90 with different versions, plus a test for "always" modifier.

tblah reviewed (approved):
Thanks for the updates. LGTM
